internal/iamauth/util.go

```diff
@@ -0,0 +1,167 @@
+	// region, err := GetRegion(configuredRegion)
+	//if err != nil {
+	//	logger.Warn(fmt.Sprintf("defaulting region to %q due to %s", DefaultRegion, err.Error()))
+	//	region = DefaultRegion
+	//}
+	cfg := aws.Config{
+		Credentials: in.Creds,
+		Region:      aws.String(in.STSRegion),
+	}
+	if in.STSEndpoint != "" {
+		cfg.Endpoint = aws.String(in.STSEndpoint)
+	} else {
+		cfg.EndpointResolver = endpoints.ResolverFunc(stsSigningResolver)
+	}
+
+	stsSession, err := session.NewSessionWithOptions(session.Options{Config: cfg})
+	if err != nil {
+		return nil, err
+	}
+
+	svc := sts.New(stsSession)
+	stsRequest, _ := svc.GetCallerIdentityRequest(nil)
+
+	// Include the iam:GetRole or iam:GetUser request in headers.
+	if in.IncludeIAMEntity {
+		entityRequest, err := formatSignedEntityRequest(svc, in)
+		if err != nil {
+			return nil, err
+		}
+
+		headersJson, err := json.Marshal(entityRequest.HTTPRequest.Header)
+		if err != nil {
+			return nil, err
+		}
+		requestBody, err := ioutil.ReadAll(entityRequest.HTTPRequest.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		stsRequest.HTTPRequest.Header.Add(in.GetEntityMethodHeader, entityRequest.HTTPRequest.Method)
+		stsRequest.HTTPRequest.Header.Add(in.GetEntityURLHeader, entityRequest.HTTPRequest.URL.String())
+		stsRequest.HTTPRequest.Header.Add(in.GetEntityHeadersHeader, string(headersJson))
+		stsRequest.HTTPRequest.Header.Add(in.GetEntityBodyHeader, string(requestBody))
+	}
+
+	// Inject the required auth header value, if supplied, and then sign the request including that header
+	if in.ServerIDHeaderValue != "" {
+		stsRequest.HTTPRequest.Header.Add(in.ServerIDHeaderName, in.ServerIDHeaderValue)
+	}
+
+	stsRequest.Sign()
+
+	// Now extract out the relevant parts of the request
+	headersJson, err := json.Marshal(stsRequest.HTTPRequest.Header)
+	if err != nil {
+		return nil, err
+	}
+	requestBody, err := ioutil.ReadAll(stsRequest.HTTPRequest.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]interface{}{
+		"iam_http_request_method": stsRequest.HTTPRequest.Method,
+		"iam_request_url":         base64.StdEncoding.EncodeToString([]byte(stsRequest.HTTPRequest.URL.String())),
+		"iam_request_headers":     base64.StdEncoding.EncodeToString(headersJson),
+		"iam_request_body":        base64.StdEncoding.EncodeToString(requestBody),
+	}, nil
+}
+
+// STS is a really weird service that used to only have global endpoints but now has regional endpoints as well.
+// For backwards compatibility, even if you request a region other than us-east-1, it'll still sign for us-east-1.
+// See, e.g., https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html#id_credentials_temp_enable-regions_writing_code
+// So we have to shim in this EndpointResolver to force it to sign for the right region
+func stsSigningResolver(service, region string, optFns ...func(*endpoints.Options)) (endpoints.ResolvedEndpoint, error) {
```

Comment:
TODO: Figure out if the [STSRegionalEndpoint](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#STSRegionalEndpoint) config option takes care of this for us.